charmcraft/config.py

```diff
@@ -14,171 +14,282 @@
-    """Take a jsonschema.ValidationError and create a proper CommandError."""
-    if error.validator == "required":
-        msg = "Bad charmcraft.yaml content; missing fields: {}.".format(
-            ", ".join(error.validator_value)
-        )
-    elif error.validator == "type":
-        expected_type = TYPE_TRANSLATOR.get(
-            error.validator_value, error.validator_value
-        )
-        field_ref = get_field_reference(error.absolute_path)
-        msg = "Bad charmcraft.yaml content; the {} must be a {}: got '{}'.".format(
-            field_ref, expected_type, error.instance.__class__.__name__
-        )
-    elif error.validator == "enum":
-        field_ref = get_field_reference(error.absolute_path)
-        msg = "Bad charmcraft.yaml content; the {} must be one of: {}.".format(
-            field_ref, ", ".join(map(repr, error.validator_value))
-        )
-    elif error.validator == "format":
-        field_ref = get_field_reference(error.absolute_path)
-        msg = "Bad charmcraft.yaml content; the {} {}: got {!r}.".format(
-            field_ref, error.cause, error.instance
-        )
-    else:
-        # safe fallback
-        msg = error.message
+    # Filter out internal __root__ detail.
+    loc = loc.replace(".__root__", "")
+    return loc
 
-    raise CommandError(msg)
 
+def format_pydantic_error_msg(msg):
+    """Format pydantic's error message field."""
+    # Replace shorthand "str" with "string".
+    msg = msg.replace("str type expected", "string type expected")
+    return msg
 
-@format_checker.checks("url", raises=ValueError)
-def check_url(value):
-    """Check that the URL has at least scheme and net location."""
-    if isinstance(value, str):
-        url = urlparse(value)
-        if url.scheme and url.netloc:
-            return True
-    raise ValueError("must be a full URL (e.g. 'https://some.server.com')")
 
+def format_pydantic_errors(errors):
+    """Format errors.
 
-@format_checker.checks("relative_path", raises=ValueError)
-def check_relative_paths(value):
-    """Check that the received paths are all valid relative ones."""
-    if isinstance(value, str):
-        # check if it's an absolute path using POSIX's '/' (not os.path.sep, as the charm's
-        # config is independent of the platform where charmcraft is running)
-        if value and value[0] != "/":
-            return True
-    raise ValueError("must be a valid relative URL")
+    Example 1: Single error.
 
+    Bad charmcraft.yaml content:
+    - field: <some field>
+      reason: <some reason>
 
-@attr.s(kw_only=True, frozen=True)
-class CharmhubConfig:
-    """Configuration for all Charmhub related options."""
+    Example 2: Multiple errors.
 
-    api_url = attr.ib(default="https://api.charmhub.io")
-    storage_url = attr.ib(default="https://storage.snapcraftcontent.com")
+    Bad charmcraft.yaml content:
+    - field: <some field>
+      reason: <some reason>
+    - field: <some field 2>
+      reason: <some reason 2>
```

Comment:
This format of "list of keys", used in yaml, may work but it's not too helpful to understand what's going on.

I'd stick to use one line per error. For example, instead of the following:

```
Bad charmcraft.yaml content:
- field: parts.bundle
  reason: value is not a valid dict
```

I prefer:

```
Bad charmcraft.yaml content:
- value is not a valid dict in field 'parts.bundle'
```

Reply:
Extra points to specify *what* is the program interpreting that is not a dict. Why not? the developer may be yelling at the screen because for they **it is a dict**.

See in opposition the current error in this situation: `Bad charmcraft.yaml content; the 'parts' field must be a dict: got 'list'.`

We really should try to get the extra mile here, otherwise we're losing functionality with this change...

Reply:
Unfortunately that information is not provided in pydantic exception and I don't see a straight forward way to capture the item that was the source of the error in pydantic.  I expect we could certainly improve some pydantic validators in the future if desired to improve context...